File: src/com/wang/Game2dEngine/action/ScaleHeightByAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action;

import static java.lang.Math.abs;

/**
 * @author ricolwang
 */
public class ScaleHeightByAction extends ScaleAction
{

    double scaleHeightBy;
    float scaleHeightByDuration;
    double scaleHeightBySpeed;
    double scaleHeightByCurrent;

    public ScaleHeightByAction()
    {
        this.scaleHeightBy = 0;
        this.scaleHeightByDuration = 0;
        this.scaleHeightBySpeed = 0;
        this.scaleHeightByCurrent = 0;
    }

    public void scaleHeightBy(double value, float duration)
    {
        if (duration <= 0)
        {
            duration = 0;
            bImmediately = true;
        }
        this.scaleHeightBy = value;
        this.scaleHeightByDuration = duration;
        if (!bImmediately)
        {
            this.scaleHeightBySpeed = value / abs(duration * 1000);
        }
        this.scaleHeightByCurrent = 0;
    }

    @Override
    public void perform(double runningTime)
    {
        if (bImmediately)
        {
            this.theSprite.setHeight(this.theSprite.getHeight() + scaleHeightBy);
            bComplete = true;
        }
        if (bComplete)
        {
            return;
        }

        if (abs(this.scaleHeightBy) > Action.MINIMUM)
        {
            double height = this.theSprite.getHeight();
            double value = this.scaleHeightBySpeed * runningTime;
            this.scaleHeightByCurrent += value;

            if (this.scaleHeightBy > 0)
            {
                if (this.scaleHeightByCurrent > this.scaleHeightBy)
                {
                    bComplete = true;
                } else
                {
                    height += value;
                }
            } else
            {
                if (this.scaleHeightByCurrent < this.scaleHeightBy)
                {
                    bComplete = true;
                } else
                {
                    height += value;
                }
            }
            this.theSprite.setHeight(height);
        } else
        {
            bComplete = true;
        }

    }
}

File: src/com/wang/Game2dEngine/action/ScaleHeightByAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

/**
 * @author ricolwang
 */
class ScaleHeightByAction : ScaleAction()
{
    var scaleHeightBy = 0.0
    var scaleHeightByDuration = 0f
    var scaleHeightBySpeed = 0.0
    var scaleHeightByCurrent = 0.0
    fun scaleHeightBy(value: Double, duration: Float)
    {
        var duration = duration
        if (duration <= 0)
        {
            duration = 0f
            bImmediately = true
        }
        scaleHeightBy = value
        scaleHeightByDuration = duration
        if (!bImmediately)
        {
            scaleHeightBySpeed = value / Math.abs(duration * 1000)
        }
        scaleHeightByCurrent = 0.0
    }

    override fun perform(runningTime: Double)
    {
        if (bImmediately)
        {
            theSprite!!.height = theSprite!!.height + scaleHeightBy
            bComplete = true
        }
        if (bComplete)
        {
            return
        }
        if (Math.abs(scaleHeightBy) > Action.Companion.MINIMUM)
        {
            var height = theSprite!!.height
            val value = scaleHeightBySpeed * runningTime
            scaleHeightByCurrent += value
            if (scaleHeightBy > 0)
            {
                if (scaleHeightByCurrent > scaleHeightBy)
                {
                    bComplete = true
                } else
                {
                    height += value
                }
            } else
            {
                if (scaleHeightByCurrent < scaleHeightBy)
                {
                    bComplete = true
                } else
                {
                    height += value
                }
            }
            theSprite!!.height = height
        } else
        {
            bComplete = true
        }
    }
}